go/v1beta1/storage/pgsqlstore_test.go

```diff
@@ -15,16 +15,375 @@
+	pgCtlStartOptions := fmt.Sprintf("-p %d", port)
+	cmd = exec.Command(pgCtl, "--pgdata", pgData.pgDataPath, "-o", pgCtlStartOptions, "start")
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+
+	pgData.startedPg = true
+
+	return nil
+}
+
+func findAvailablePort() (availablePort int, err error) {
+	for availablePort = 5432; availablePort < 6000; availablePort++ {
+		l, err := net.Listen("tcp", fmt.Sprintf(":%d", availablePort))
+		defer l.Close()
+		if err == nil {
+			return availablePort, nil
+		}
+	}
+
+	return -1, fmt.Errorf("Unable to find an open port")
+}
+
+func isPostgresRunning(config *config.PgSQLConfig) bool {
+	source := storage.CreateSourceString(config.User, config.Password, config.Host, "postgres", config.SSLMode)
+	db, err := sql.Open("postgres", source)
+	if err != nil {
+		return false
+	}
+	defer db.Close()
+
+	if db.Ping() != nil {
+		return false
+	}
+	return true
+}
+
+func getPostgresBinPathFromSystemPath() (binPath string, err error) {
+	cmd := exec.Command("which", "pg_ctl")
+	output, err := cmd.Output()
+	if output != nil && err == nil {
+		binPath = filepath.Dir(string(output))
+	}
+	return
+}
+
+func setup() (pgData *testPgHelper, err error) {
+	pgConfig := &config.PgSQLConfig{
+		Host:     "127.0.0.1:5432",
+		User:     "postgres",
+		Password: "password",
+		SSLMode:  "disable",
+	}
+
+	pgData = &testPgHelper{
+		startedPg: false,
+		pgConfig:  pgConfig,
+	}
+
+	//See if postgres is already available and running
+	if isPostgresRunning(pgConfig) {
+		return
+	}
+
+	workingDir, err := os.Getwd()
+	if err != nil {
+		return pgData, err
+	}
+	//The root is 3 levels up. Not an ideal means of getting this, but since there are multiple
+	//ways of starting the tests, os.Executable and runtime.Caller don't seem like viable options
+	basepath := filepath.Dir(filepath.Dir(filepath.Dir(workingDir)))
+	basepath = filepath.Join(basepath, "test", "utilities")
+
+	pgData.pgBinPath = ""
+	testPgPath := filepath.Join(basepath, "pgsql")
+	testPgBinPath := filepath.Join(testPgPath, "bin")
+	if _, err := os.Stat(testPgBinPath); !os.IsNotExist(err) {
+		pgData.pgBinPath = testPgBinPath
+	} else {
+		//Check for a global installation
+		globalPgBinPath, err := getPostgresBinPathFromSystemPath()
+		if err == nil && globalPgBinPath != "" {
+			pgData.pgBinPath = globalPgBinPath
+		}
+	}
+
+	if pgData.pgBinPath == "" {
+		os.MkdirAll(basepath, 0755)
+
+		fileNameBase := "postgresql-10.15.1"
+		postgresZipFileName := filepath.Join(basepath, fmt.Sprintf("%s.zip", fileNameBase))
+
+		postgresZipFound := false
+		if _, err := os.Stat(postgresZipFileName); !os.IsNotExist(err) {
+			postgresZipFound = true
+		}
+
+		if !postgresZipFound {
+			//URLs from https://www.enterprisedb.com/download-postgresql-binaries
+			postgresUrl := ""
+			switch runtime.GOOS {
+			case "darwin":
+				postgresUrl = "https://sbp.enterprisedb.com/getfile.jsp?fileid=1257418&_ga=2.59164973.1284458936.1605634932-924989421.1605634932"
```

Comment:
Any reason why this files gets downloaded from EnterpriseDB
I am not sure that would be such a good idea

Reply:
Nothing nefarious was meant. Just got the download links from https://www.postgresql.org/download/windows/ → [zip archive](https://www.postgresql.org/download/windows/#pgContentWrap:~:text=Advanced%20users%20can%20also%20download%20a%20zip%20archive%20of%20the%20binaries%2C%20without%20the%20installer). The intent was to have a locally accessible binary without requiring developers to do a full PostgreSQL installation.

If there are other links that you'd recommend, I'm totally open to suggestions.

Reply:
Thanks for the reply I think the direct links are way better